lib/dslash_domain_wall_4d_fused_m5.hpp

```diff
@@ -132,23 +123,44 @@ namespace quda
   template <typename Float, int nColor, QudaReconstructType recon> struct DomainWall4DApplyFusedM5 {
 
     template <Dslash5Type dslash5_type_impl, Dslash5Type... N>
-    inline DomainWall4DApplyFusedM5(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, double a,
-                                    double m_5, const Complex *b_5, const Complex *c_5, const ColorSpinorField &x,
-                                    ColorSpinorField &y, int parity, bool dagger, const int *comm_override, double m_f,
-                                    Dslash5TypeList<dslash5_type_impl, N...>, TimeProfile &profile)
+    DomainWall4DApplyFusedM5(cvector_ref<ColorSpinorField> &out, cvector_ref<const ColorSpinorField> &in,
+                             cvector_ref<const ColorSpinorField> &x, const GaugeField &U, cvector_ref<ColorSpinorField> &y,
+                             const Complex *b_5, const Complex *c_5, double a, double m_5, int parity,
+                             bool dagger, const int *comm_override, double m_f,
+                             Dslash5TypeList<dslash5_type_impl, N...>, TimeProfile &profile)
     {
 #ifdef NVSHMEM_COMMS
       errorQuda("Fused Mobius/DWF-4D kernels do not currently work with NVSHMEM.");
 #else
       constexpr int nDim = 4;
+      auto halo = ColorSpinorField::create_comms_batch(in);
       using Arg = DomainWall4DFusedM5Arg<Float, nColor, nDim, recon, dslash5_type_impl>;
-      Arg arg(out, in, U, a, m_5, b_5, c_5, a != 0.0, x, y, parity, dagger, comm_override, m_f);
-      DomainWall4DFusedM5<Arg> dwf(arg, out, in);
-
-      dslash::DslashPolicyTune<decltype(dwf)> policy(dwf, in, in.getDslashConstant().volume_4d_cb,
-                                                     in.getDslashConstant().ghostFaceCB, profile);
+      Arg arg(out, in, halo, U, a, m_5, b_5, c_5, a != 0.0, x, y, parity, dagger, comm_override, m_f);
+      DomainWall4DFusedM5<Arg> dwf(arg, out, in, halo, y);
+      dslash::DslashPolicyTune<decltype(dwf)> policy(dwf, in, halo, profile);
 #endif
     }
   };
 
+  // use custom instantiate to deal with field splitting if needed
+  template <template <typename, int, QudaReconstructType> class Apply, typename Recon = WilsonReconstruct, typename... Args>
```

Comment:
You could create an `instantiate_recurse4` function for this one---I know we can't chase recursion counts forever, but I think it's worth it since this is still a dslash, and it's best to have a single file where all of the dslash recursive instantiations live.

Reply:
The issue with doing that is that with it being a dslash, it uses the instantiate functions in `instantiate_dslash.h` not `instantiate.h`.

Reply:
Noting this can be done however in the future when I delete the separate `instantiate_dslash.h` header.